[V1] [P/D] Add Support for KV Load Failure Recovery

sdavidbd opened the pull request:
## 🎯 Purpose
This PR implements the design proposed in [RFC #19329](https://github.com/vllm-project/vllm/issues/19329).

It adds a mechanism to **recover from KV load failures** in vLLM’s KV connector path by:
- Detecting failed KV block loads
- Automatically rescheduling affected requests for recomputation from a valid prefix

The implementation supports both **synchronous** and **asynchronous** loading methods, improving robustness when using external systems for KV cache offload or transfer.

---

## 🧪 Test Plan
### ✅ Unit Tests
#### Coverage:
- Recovery logic for sync and async KV load failures
- Shared block scenarios (only applicable to sync loading)
- Correct request rescheduling and computed-token count adaptation
#### Run:
```bash
pytest -v tests/v1/kv_connector/unit/test_output_aggreagator.py
pytest -v tests/v1/kv_connector/unit/test_kv_load_failure_recovery.py
```
#### Results:
```
✔ test_output_aggreagator.py
  ├─ test_aggregate_workers_output
  └─ test_async_aggregate_workers_output

✔ test_kv_load_failure_recovery.py
  ├─ test_async_load_failure (3 cases)
  ├─ test_sync_load_failure (3 cases)
  ├─ test_sync_load_failure_with_shared_blocks (3 cases)
  └─ test_async_progressive_load_failure (2 cases)
```
### 📦 End-to-End Example
A full integration test is available at:
```bash
examples/offline_inference/kv_load_failure_recovery/
```
This test demonstrates the recovery mechanism using fault-injecting connectors that simulate failed KV loads.
Test flow:

1. Prefill stage — Saves KV data to the local filesystem
2. Decode stage 1 — Successful sync KV load (baseline)
3. Decode stage 2 — Failed sync KV load
4. Decode stage 3 — Failed async KV load

See the [README.md](https://github.com/vllm-project/vllm/pull/examples/offline_inference/kv_load_failure_recovery/README.md) for details.
#### Run:
```bash
cd examples/offline_inference/kv_load_failure_recovery
./run.sh
```
#### Results:
All outputs in simulated failure cases match the baseline decode.
```
...
INFO:root:Simulating failure to load all KV blocks for the first load request. Total blocks: 62
...
INFO 08-07 17:21:21 [scheduler.py:1037] Recovered from KV load failure: 2 request(s) rescheduled (1022 tokens affected).
...
INFO:root:Simulating failure to load all KV blocks for the first load request. Total blocks: 62
...
INFO 08-07 17:21:39 [scheduler.py:1037] Recovered from KV load failure: 1 request(s) rescheduled (0 tokens affected).
...
✅ Outputs match: recovery successful.
```

Commit:
Add KV load failure recovery by rescheduling affected requests

Enable graceful recovery from KV connector V1 load errors. If external
KV blocks fail to load, the connector reports their IDs, and the
Scheduler automatically reschedules requests referencing them. Affected
tokens are recomputed from a valid prefix.

Signed-off-by: David Ben-David <davidb@pliops.com>

sdavidbd commented on vllm/v1/core/sched/scheduler.py:
```diff
@@ -723,6 +726,25 @@ def update_from_output(
+            if invalid_block_ids:
+                reschedule_request = False
+                req_block_ids = self.kv_cache_manager.get_block_ids(req_id)[0]
+                for idx, block_id in enumerate(req_block_ids):
+                    if idx >= request.num_computed_tokens:
+                        break
+                    if block_id in invalid_block_ids:
+                        num_requests_to_reschedule += 1
+                        num_tokens_to_reschedule += request.num_computed_tokens
+                        request.num_computed_tokens = idx * self.block_size
+                        num_tokens_to_reschedule -= request.num_computed_tokens
+                        new_running.append(request)
+                        reschedule_request = True
+                        break
```
Good catch — the intention behind the condition was to exit early and avoid unnecessary iteration, but due to the token/block mismatch, it doesn’t work as intended. While this has no impact on correctness (it won’t cause premature termination), it may lead to some redundant work in some cases. I’ll fix the condition to behave as expected. The suggested change is valid, but I prefer keeping enumerate for clarity and performance, so I’ll adapt the fix accordingly.

sdavidbd commented on examples/offline_inference/kv_load_failure_recovery/decode_example.py:
```diff
@@ -0,0 +1,70 @@
+    )
+    args = parser.parse_args()
+
+    if args.simulate_failure:
+        ktc = KVTransferConfig(
+            kv_connector="RogueSharedStorageConnector",
+            kv_role="kv_both",
+            kv_connector_extra_config={
+                "shared_storage_path": "local_storage",
+            },
+            kv_connector_module_path="rogue_shared_storage_connector",
+        )
+    else:
+        ktc = KVTransferConfig(
+            kv_connector="SharedStorageConnector",
+            kv_role="kv_both",
+            kv_connector_extra_config={
+                "shared_storage_path": "local_storage",
+            },
+        )
```
In general I agree that reduced duplication improves readability and maintainability, but in the context of this **test code**, I think the original form may better serve readability as each branch is fully self-contained and easy to understand without indirection. Also, the shared configuration in not likely to change.

orozery commented:
@sdavidbd the problem I see here is in multi-worker case.
For example, if you use tensor-parallelism or pipeline parallelism.
Each worker may fail loading on different block ids.
Although each worker reports its own invalid_block_ids, the scheduler will only get the invalid_block_ids of the first-rank worker.
This is due to the implementation of `MultiprocExecutor.execute_model` which simply discards the `ModelRunnerOutput` of all but the first worker.

In theory, the same problem holds for the existing fields `ModelRunnerOutput.finished_sending` and `ModelRunnerOutput.finished_recving`.
The current solution in the code for this is having the first-rank worker communicate with all other workers to aggregate input from all workers to its own output.
This is what was done in `NixlConnector.get_finished`.

However, I this it is better to have aggregation by the scheduler, and not by the workers.
And also, have one aggregation per all fields of `ModelRunnerOutput`.

I'm actually planning on opening up a PR in the upcoming days that will add a generic `KVConnectorMetadata` field to `ModelRunnerOutput`. This field will be used to transfer ALL required connector information from workers to the scheduler. This includes `finished_sending` and `finished_recving`, and potentially your `invalid_block_ids`.

With this approach, the `get_block_ids_with_load_errors` will become a scheduler-side connector call and will extract the `invalid_block_ids` from the abstract `KVConnectorMetadata`.

My motivation for this is my plan to add an offloading connector, and be able to report offloading status from workers to the scheduler.

cc @njhill

sdavidbd commented:
Converting this PR to draft for now - I'm working on extending support for tensor-parallel (TP) setups and adding more unit tests to improve coverage and reliability. Will mark as ready for review once that's complete.

njhill commented:
Thanks @sdavidbd, and sorry for the delay getting back to this.

I guess I'm still not convinced that block ids would be "simpler" - since we already pass back ids of finished requests, just having these flagged as failed or succeeded seems simpler to me, and would be straightfoward on the scheduler side for example to just fall back to locally prefilling any failed requests.

Since errors would presumably be quite rare, the granularity optimization seems like it might be unnecessary.

One thing I have been wondering about related to this though, is how we should we handle cases where concurrent requests want to load overlapping prefixes. If there is an async kv load in already flight for a subset of the blocks, the request should ideally just subscribe / wait behind that (in addition to loading additional blocks in parallel, if needed). I don't think that's how it currently works.

NickLucche commented:
Thanks for the work @sdavidbd !

re: @orozery 
> I'm actually planning on opening up a PR in the upcoming days that will add a generic KVConnectorMetadata field to ModelRunnerOutput.

This is actually independent from having the reduce carried out in the scheduler though.
We could still have rank0 be in charge of aggregating data from the other ranks, so that the actual logic remains where the data is produced. Perhaps I am just missing the advantage of scheduler-side reduction.
But we definitely need to aggregate as you pointed out.

> having these (request_id) flagged as failed or succeeded seems simpler to me

I agree here I think it would make the aggregation above simpler and would move less data around the ranks. I believe we can overload the single communication step where we already exchange finished_req_ids.

sdavidbd commented:
Thanks @njhill - really appreciate the follow-up.

You're absolutely right that today errors are relatively rare. However, as **KV cache offloading** gains traction - especially with the rise of disaggregated storage solutions - we do expect failures to become more common. Failures could stem from transient disconnections or external eviction policies, and designing for them upfront helps ensure resilience.

Looking forward, KV cache offloading is expected to play a key role in enabling KV reuse across various scenarios: preempted requests, multi-turn conversations, long shared document queries, and more. In these cases, externally computed tokens are loaded as part of **running** prefill requests and can potentially be shared by multiple requests in the same batch.

Critically, the connector typically does not have full visibility into all requests that might share these blocks - specifically, requests beyond the first will treat the blocks as if they were computed locally rather than loaded externally. Reporting block-level failures allows the scheduler, which does maintain a global view of request-to-block mappings, to identify and reschedule only the affected requests efficiently.

Moreover, the future enhancement I mentioned - supporting reuse of arbitrary prompt blocks rather than only the prefix - is an important and feasible optimization we are actively working on. The block-level failure reporting design directly aligns with and enables this finer-grained reuse and recovery path.

Regarding your point about overlapping async KV loads: I agree - currently, overlapping blocks are duplicated locally instead of being shared. While this gap is certainly addressable, it also raises a broader question:
Why do disaggregated decode requests need to fully wait for remotely computed KV blocks to finish loading before starting at all? Couldn't we schedule the requests immediately and load the KV blocks asynchronously, layer by layer, during the forward pass?
This approach could resolve the duplication issue and allow us to overlap block loading with computation, further optimizing end-to-end latency.

Thanks again for the constructive discussion - happy to further align on this if you'd like!

njhill commented:
Thanks @sdavidbd. Maybe it's better to have the connector API be block-based? (see discussion in other PR https://github.com/vllm-project/vllm/pull/19555#discussion_r2187014032).

> Couldn't we schedule the requests immediately and load the KV blocks asynchronously, layer by layer, during the forward pass?

This is already supported by the connector API but there may be a trade-off since we need to load at least the first layer before the forward pass starts, which will add latency to every other request in the batch. It might also slow down subsequent layer if the loading takes longer than computation of the prior layer. There might also be greater overhead associated with multiple per-layer transfers. But it's something we could experiment with. We want to at least support async for the initial handshake.

I guess we need to think through the sync and async loading cases with respect to block sharing. You're right that the sharing happens automatically/implicitly in the sync case right now, it would be good to have that work for the async case too.

Commit:
Add KV load failure recovery example

Signed-off-by: David Ben-David <davidb@pliops.com>

ApostaC commented:
Thanks @sdavidbd ! I like the idea of this PR. This will also be very useful for LMCache's use cases.

orozery commented on vllm/distributed/kv_transfer/kv_connector/v1/base.py:
```diff
@@ -206,6 +206,16 @@ def get_finished(
         """
         return None, None
 
+    def get_block_ids_with_load_errors(self) -> Optional[set[int]]:
```
IIRC @njhill suggested that we move the connector to be block-based instead of request-based.
I was thinking we can change `finished_sending` and `finished_recving` to be block-based, and then we don't need this new `get_block_ids_with_load_errors` API.
This will also align better with my plans for a block-based offloading connector.

@sdavidbd What do you think?

sdavidbd replied:
> IIRC @njhill suggested that we move the connector to be block-based instead of request-based. I was thinking we can change `finished_sending` and `finished_recving` to be block-based, and then we don't need this new `get_block_ids_with_load_errors` API. This will also align better with my plans for a block-based offloading connector.
> 
> @sdavidbd What do you think?

I strongly believe that adopting a block-level connector design is a cleaner and more general approach. It also offers a more elegant alternative to the `invalid_block_ids` set—by associating a success/failure flag with each "finished" block.

With that said, since this represents a more substantial change and may require broader alignment, it might be best to defer it to a future PR.

yoo-kumaneko commented:
This feature is critically important. Without a failure handling mechanism, any cache storage system is essentially unusable.

sdavidbd commented:
- Rebased the PR and resolved merge conflicts.
- Refactored `GpuModelRunner.execute_model` to encapsulate the KV connector lifecycle using a context manager.  
  This simplifies propagating additional data back from the KV connector and ensures `clear_connector_metadata` is properly called.  
  _(The refactor can be extracted to a separate PR if it needs to be merged independently.)_

**Currently WIP in this PR:**
- Add support for async KV load failure recovery
- Implement `get_block_ids_with_load_errors` API in `MultiConnector`

**ETA:** July 27

sdavidbd commented:
Added failure recovery support for asynchronous KV load. Covered by the system test in `examples/offline_inference/kv_load_failure_recovery`.
Additional unit tests will be added in a follow-up commit.

njhill reviewed (commented):
Thanks again @sdavidbd I think the code looks great.

> Refactored `GpuModelRunner.execute_model` to encapsulate the KV connector lifecycle using a context manager.
This simplifies propagating additional data back from the KV connector and ensures `clear_connector_metadata` is properly called.

I like this but as you suggested given it's orthogonal it may be better to move it to a dedicated PR which we can hopefully then get merged much quicker :sweat_smile:

sdavidbd commented:
> Thanks again @sdavidbd I think the code looks great.
> 
> > Refactored `GpuModelRunner.execute_model` to encapsulate the KV connector lifecycle using a context manager.
> > This simplifies propagating additional data back from the KV connector and ensures `clear_connector_metadata` is properly called.
> 
> I like this but as you suggested given it's orthogonal it may be better to move it to a dedicated PR which we can hopefully then get merged much quicker 😅

Moved the refactor into a separate PR: #21980

liuzijing2014 commented:
Hi @sdavidbd, thanks for the PR. Something additionally, for KV connectors (NIXLConnector) who use async API (e.g. `get_finished`), there are cases that KV fetching from remote keeps failing, and in that case, we want to notify `scheduler` to abort/finish/remove that request from further processing and free up its blocks. The current `get_finished` API only notifies `scheduler` which request is finished, but no way to tell which request can't be finished at all.

Do you think we can also cover this case as well in this PR?

Commit:
Add KV load failure recovery unit tests

Signed-off-by: David Ben-David <davidb@pliops.com>

Commit:
Add failure recovery support for async KV load

Signed-off-by: David Ben-David <davidb@pliops.com>

liuzijing2014 commented:
> First, just to clarify: this PR focuses on adding the infrastructure for automatic recovery on the _decoder side_ in the event of KV load failures. Specifically, when a connector reports failed block loads via `get_block_ids_with_load_errors`, the decoder can fall back to recomputing the missing blocks. However, this PR does not yet include changes to `NixlConnector` to support this mechanism (i.e., it doesn't implement `get_block_ids_with_load_errors`), so failure recovery is not currently enabled for that use case. That will be handled in a follow-up PR.
> 
> As for your question about the _prefiller_ side: you're right that if a remote load fails, the connector may not report the request as finished - which would prevent the scheduler from releasing the associated blocks. To handle this, we can rely on the timeout mechanism introduced in #20139, which ensures that such requests are eventually cleaned up.

Thanks for the explanation @sdavidbd! It sounds like this PR covers the error handling for KV connectors who do KV injection/load "in-place":

`start_load_kv` on request 1 -> `wait_kv_load` on request 1 layer 1 -> attn run for layer 1 -> `wait_kv_load` on request 1 layer 2 -> attn run for layer 2 -> and so on ...

I am more referring to the case where KV connectors who do KV injection/load fully async regarding to model execution where a request would only be scheduled/batched for model run (forward execution) after its KV loading is fully complete on all layers.

`start_load_kv` on request 1 to trigger fully async where request 1 is not even scheduled to run in the current model forward iteration -> `wait_kv_load` no ops -> attn run for layer 1 -> `wait_kv_load` no ops -> attn run for layer 2 -> so on .. -> `get_finished` informs scheduler request KV injection/loading is complete -> request 1 would be scheduled for model run for the next iteration.

I think `NIXLKVConnector` and our vendor internal KV connector solution both adopt this approach.

sdavidbd commented:
Updated PR description to highlight support for both sync and async load handling, and to detail corresponding test coverage.

liuzijing2014 commented:
Get it @sdavidbd. Just to double confirm: in the case of fully async, when `scheduler` handles a failed request (whose kv loading has failed), it would reschedule such request to `start_load_kv` in the next round again?

sdavidbd commented:
Hi all — I know there are ongoing discussions about the longer-term KV-connector API. This PR is intentionally **orthogonal and backward-compatible** (it only adds an API and doesn’t change existing behavior) and it addresses an immediate gap: **robust failure recovery for both synchronous and asynchronous KV loading.**

#### Why now
- Production users are already hitting KV-load failures; this provides a safe, minimal remedy.
- No performance regressions on the happy path — logic only activates on failure.

#### What’s in
- Block-level failure detection + rescheduling from a valid prefix (additive interface changes).
- Support for both sync and async loading paths.
- Comprehensive unit + system tests included.

All prior feedback has been addressed.
Could a reviewer with write access provide a final approval, or clarify the one or two items still blocking? I’ll handle them promptly. 🙏

@WoosukKwon @njhill @KuntaiDu

sdavidbd commented on vllm/v1/core/sched/scheduler.py:
```diff
@@ -923,6 +936,121 @@ def update_from_output(
+    def _update_requests_with_invalid_blocks(
+            self, requests: Iterable[Request],
+            invalid_block_ids: set[int]) -> tuple[set[Request], int, set[int]]:
+        affected_requests: set[Request] = set()
```
Thanks @lk-chen — this implementation is focused on **automatic recovery** from KV load failures by falling back to recomputing the blocks that failed to load. Since inference is time-critical, I’m not sure a retry mechanism is appropriate: retries will increase latency without strong guarantees of success, while recomputation is always safe and bounded in cost.

For the **asynchronous loading path**, if limited retries are desirable, they can be fully handled inside the connector implementation itself — the connector would only report a block as “failed” once it has exhausted all retry attempts. That way, the scheduler remains simple and only reacts to definitive failures.

njhill replied:
In the async case we might want to have the retries handled by the framework. So that successful blocks are eagerly cached, and other connectors get a chance to provide the cache in multi-connector case.

sdavidbd replied:
Following my last fix for handling async KV load failures, this logic now runs in `_update_from_kv_xfer_finished`, after async loading has completed (see my [other comment](https://github.com/vllm-project/vllm/pull/19330#discussion_r2328878457)). At that point, I don’t think the extra overhead of freeing the blocks is justified by the relatively small chance of a cache hit in other connectors.

njhill commented on vllm/v1/core/sched/scheduler.py:
```diff
@@ -923,6 +936,121 @@ def update_from_output(
+            else:
+                # No valid computed tokens, release allocated blocks.
+                # There may be a local cache hit on retry.
+                self.kv_cache_manager.free(request)
```
I think we need to free in both cases, since the request will be re-allocated? Any new cached blocks will be hit and connectors will have a chance to retrieve any remaining tokens.

njhill commented on vllm/v1/core/sched/scheduler.py:
```diff
@@ -923,6 +936,121 @@ def update_from_output(
+        total_requests_to_reschedule += len(affected_requests)
+        total_tokens_to_reschedule += num_tokens_to_reschedule
+
+        for request in affected_requests:
+            if request.num_computed_tokens:
+                # Cache any valid computed tokens.
+                self.kv_cache_manager.cache_blocks(request,
+                                                   request.num_computed_tokens)
+            else:
+                # No valid computed tokens, release allocated blocks.
+                # There may be a local cache hit on retry.
+                self.kv_cache_manager.free(request)
+
+            request.status = RequestStatus.WAITING
```
I think we need to look more closely here at the interaction of this with the `finished_recving` notifications.

We can probably assume that the invalid block notification will come through from the connector in the same (or possibly earlier) step as that request's id in the `finished_recving` set.

But that gets processed later in this method (in `_update_from_kv_xfer_finished`), with those reqs added to the `self.finished_recving_kv_req_ids` set. It is assumed that req ids in that set will correspond to reqs in `WAITING_FOR_REMOTE_KVS` state, since these are checked by the call to `_update_waiting_for_remote_kv` in the subsequent `schedule()`. So now the entries in that set will be orphaned or worse mis-interpreted when the request is re-scheduled.

I haven't thought this through completely but I think we probably need to move part of this logic to `_update_waiting_for_remote_kv` ... when it processes the finished req id it should obtain the corresponding list of invalid block ids available and take that into account.

sdavidbd replied:
I think the right approach is to add a request to `finished_recving_kv_req_ids` only if it’s still in the `WAITING_FOR_REMOTE_KVS` state. If a request waiting on async load encounters a failed block, we move it back to `WAITING` before handling the `finished_recving` set — so there shouldn’t be any issues.

sdavidbd replied:
@njhill I thought it over again, and you’re right. With **progressive reporting** of failed blocks, we must not move a request from `WAITING_FOR_REMOTE_KVS` to `WAITING` before the async load actually completes; doing so risks a race between the connector (still transferring blocks) and the attention kernel (reusing blocks).

I changed the logic so that:
- In `_update_requests_with_invalid_blocks`, we process failed async KV loads by updating `num_computed_tokens` and leaving the request in `WAITING_FOR_REMOTE_KVS`. We also record it in `failed_recving_kv_req_ids`.
- Then in `_update_from_kv_xfer_finished` (called per request), we check whether the request is in `failed_recving_kv_req_ids` and skip re-setting `num_computed_tokens` in that case.

This keeps the scheduling assumptions intact, aligns with progressive block-level reporting, and prevents “orphaned” or mis-interpreted entries in `finished_recving_kv_req_ids`.

njhill commented on vllm/v1/core/sched/scheduler.py:
```diff
@@ -923,6 +936,121 @@ def update_from_output(
+        return (affected_requests, num_tokens_to_reschedule,
+                marked_invalid_block_ids)
+
+    def _handle_invalid_blocks(self, invalid_block_ids: set[int]) -> set[str]:
```
I think this should be moved to the `KV Connector Related Methods` section at the bottom of this file. It can take `KVConnectorOutput` arg as input...

njhill commented on vllm/v1/core/sched/scheduler.py:
```diff
@@ -923,6 +936,121 @@ def update_from_output(
 
         return engine_core_outputs
 
+    def _update_requests_with_invalid_blocks(
```
Could this be moved to `KV Connector Related Methods` section at the bottom of this file

njhill reviewed (commented):
Thanks @sdavidbd and sorry again for the delay with this.

Some general comments/questions in addition to the inline ones:
- I think [`MultiConnector`](https://github.com/vllm-project/vllm/blob/main/vllm/distributed/kv_transfer/kv_connector/v1/multi_connector.py) needs updating with the new method
- What changes will be needed when we incorporate https://github.com/vllm-project/vllm/pull/23624? Probably the API will need to change from int block ids?
- How will this work with async scheduling? In particular perhaps you could take a look at https://github.com/vllm-project/vllm/pull/23569 which we're planning to incorporate soon

njhill commented on vllm/v1/core/sched/scheduler.py:
```diff
@@ -775,17 +775,30 @@ def update_from_output(
         num_scheduled_tokens = scheduler_output.num_scheduled_tokens
         pooler_outputs = model_runner_output.pooler_output
         num_nans_in_logits = model_runner_output.num_nans_in_logits
+        invalid_block_ids = (
+            model_runner_output.kv_connector_output.invalid_block_ids
+            if model_runner_output.kv_connector_output else None)
```
Might not be relevant with the other suggested refactor though

njhill commented:
@sdavidbd I meant to also say that in general the code is very nicely written!

Also after thinking through the details, I feel like we should ideally include support for progressive completion reporting for the async case. Since we would now be reporting failures at the block level, it would be good to somehow indicate which blocks have succeeded so far as well as which have failed (this would also extend to saving as well as loading) - i.e. even thought the entire request might not yet be finished at that step. For this, it may be cleaner to communicate a count of tokens or blocks rather than a long list.

Commit:
Fix unit tests after rebase

Signed-off-by: David Ben-David <davidb@pliops.com>

Commit:
Refactor scheduler changes

Signed-off-by: David Ben-David <davidb@pliops.com>

Commit:
Fix and refactor rollback handling of invalid sampled tokens

Signed-off-by: David Ben-David <davidb@pliops.com>

Commit:
Refactor: always return a set of failed block IDs from connectors

Connectors now consistently return a set of failed block IDs (empty if none).
This removes the need for Optional handling, simplifying the code without
introducing any performance impact.

Signed-off-by: David Ben-David <davidb@pliops.com>

Commit:
Implement get_block_ids_with_load_errors in MultiConnector

Signed-off-by: David Ben-David <davidb@pliops.com>

Commit:
Emphasize that KV load failure recovery does not currently support HMA

Signed-off-by: David Ben-David <davidb@pliops.com>

sdavidbd commented:
> @sdavidbd I meant to also say that in general the code is very nicely written!
> 
> Also after thinking through the details, I feel like we should ideally include support for progressive completion reporting for the async case. Since we would now be reporting failures at the block level, it would be good to somehow indicate which blocks have succeeded so far as well as which have failed (this would also extend to saving as well as loading) - i.e. even thought the entire request might not yet be finished at that step. For this, it may be cleaner to communicate a count of tokens or blocks rather than a long list.

Thank you, Nick — I really appreciate the feedback. Your suggestion fully aligns with my vision for a block-based connector. That said, I don’t think this can be handled reliably by just communicating counts. Progressive loading doesn’t necessarily guarantee that blocks are completed in order. Reporting block IDs directly ensures correctness, simplifies handling, and leaves room for future enhancements. We can safely assume that each block will be reported exactly once, when its load has either succeeded or failed.

Commit:
Fix handling of async KV load failures

Signed-off-by: David Ben-David <davidb@pliops.com>

sdavidbd commented:
@njhill I’ve pushed fixes addressing all review comments.

This PR now supports **progressive reporting of async KV load failures**. Recovery awaits completion of async loading, which is currently based on the connector reporting the request in `finished_recving`. In future work this will be replaced by progressive reporting of successfully loaded blocks (orthogonal to this PR).

For now, **HMA** and **async scheduling** remain unsupported and will be addressed in follow-up PRs.

Commit:
Fix handling of progressive async KV load failures

Signed-off-by: David Ben-David <davidb@pliops.com>

ApostaC commented on vllm/v1/core/sched/scheduler.py:
```diff
@@ -1246,18 +1258,33 @@ def _update_waiting_for_remote_kv(self, request: Request) -> bool:
+                self.kv_cache_manager.cache_blocks(request,
+                                                   request.num_computed_tokens)
+            else:
+                # No valid computed tokens, release allocated blocks.
+                # There may be a local cache hit on retry.
+                self.kv_cache_manager.free(request)
```
In the current scheduler, when `request.num_computed_tokens == 0`, it will go through the connector lookup again: https://github.com/vllm-project/vllm/blob/81dea2077ede2d3b7e9348558c9812b9b3199108/vllm/v1/core/sched/scheduler.py#L379-L390

Given that the connector already failed to retrieve the KV cache for this request, going through the connector API again is likely to have the same failure and cause a "deadloop" for this request.

Do we want to have a logic in the scheduler to prevent such a deadloop? Or we make it explicit that the connector needs to handle this?

ApostaC replied:
Thanks for the clarification! Please leave some docstring in the base class to make it explicit 🙏. This can potentially help a lot when other contributors are debugging

ApostaC commented on vllm/v1/core/sched/scheduler.py:
```diff
@@ -1285,3 +1312,114 @@ def _update_from_kv_xfer_finished(self,
+            invalid_block_ids: set[int]) -> tuple[set[str], int, set[int]]:
+        affected_req_ids: set[str] = set()
+        total_affected_tokens = 0
+        # If a block is invalid and shared by multiple requests in the batch,
+        # all requests must be rescheduled, but only the first will recompute
+        # it. This set tracks blocks already marked for recomputation.
```
Will "a block is invalid and shared by multiple requests in the batch" really happen?

@sdavidbd @njhill Correct me if I'm wrong, there should be some logic marking the connector-loading blocks not sharable (i.e., not cached).

sdavidbd replied:
I agree with @njhill — the `kv_load_failure_recovery` example in this PR demonstrates this clearly. In both sync and async loading cases, I used two pairs of prompts with a shared prefix. When a KV load error occurs in the sync case, it impacts both requests in the pair. In the async case, however, only one request is affected, and when that request is rescheduled it successfully reuses the local cache created by the other request that completed loading.

As I noted in previous comments on this PR, supporting block sharing for the async case should be relatively straightforward, and I’d be happy to contribute that in a follow-up PR.

ApostaC replied:
Ah, this makes sense.

ApostaC commented on vllm/v1/core/sched/scheduler.py:
```diff
@@ -1285,3 +1312,114 @@ def _update_from_kv_xfer_finished(self,
+        # all requests must be rescheduled, but only the first will recompute
+        # it. This set tracks blocks already marked for recomputation.
+        marked_invalid_block_ids: set[int] = set()
+        for request in requests:
```
The nested loop will go through the blocks of all the requests in `running` and `WAITING_FOR_REMOTE_KVS` state, which may cause non-negligible overheads. (Especially when the batch size is as large as hundreds of requests)

ApostaC commented on vllm/v1/core/sched/scheduler.py:
```diff
@@ -1285,3 +1312,114 @@ def _update_from_kv_xfer_finished(self,
+                    continue
+
+                marked_invalid_block = True
+                total_affected_tokens += req_num_computed_tokens
+                request.num_computed_tokens = idx * self.block_size
+                total_affected_tokens -= request.num_computed_tokens
```
nit:
```suggestion
                request.num_computed_tokens = idx * self.block_size
                total_affected_tokens += req_num_computed_tokens - request.num_computed_tokens
```

ApostaC commented on vllm/v1/core/sched/scheduler.py:
```diff
@@ -1285,3 +1312,114 @@ def _update_from_kv_xfer_finished(self,
+
+                affected_req_ids.add(request.request_id)
+
+        return (affected_req_ids, total_affected_tokens,
+                marked_invalid_block_ids)
```
The `total_affected_tokens` is only used for logging at the caller. Does this function really need to return such a value (at the cost of increasing the mental complexity to understand the code)?

ApostaC commented on vllm/v1/core/sched/scheduler.py:
```diff
@@ -870,17 +871,29 @@ def update_from_output(
         outputs: dict[int, list[EngineCoreOutput]] = defaultdict(list)
         spec_decoding_stats: Optional[SpecDecodingStats] = None
 
+        recovered_req_ids = None
```
nit: `recovered_req_ids` is a bit confusing -- we can use `req_ids_with_failure` (explicitly expressing there are failures in this request) or `affected_req_ids` (matching the variable name used in `_handle_invalid_blocks`.

ApostaC reviewed (commented):
Did a pass on part of the PR (mostly the scheduler-side changes).
High-level comment/question: Is there any reason we want to stick to the block-based error reporting? 
- Right now, the "successful KV loading" is reported at the request-level, which is easy to follow. 
- However, the current block-based error reporting needs to convert the failed block ids to the request ids and re-calculate the `num_computed_tokens` in the scheduler, which incurs a lot of mental complexity to understand the code.

sdavidbd commented:
> Did a pass on part of the PR (mostly the scheduler-side changes). High-level comment/question: Is there any reason we want to stick to the block-based error reporting?
> 
> * Right now, the "successful KV loading" is reported at the request-level, which is easy to follow.
> * However, the current block-based error reporting needs to convert the failed block ids to the request ids and re-calculate the `num_computed_tokens` in the scheduler, which incurs a lot of mental complexity to understand the code.

@ApostaC Thank you for the review.

Block-based error reporting is intentional and aligns with the broader vision of a block-based KV connector. This has been discussed both in this PR (e.g., see [this comment](https://github.com/vllm-project/vllm/pull/19330#issuecomment-3139561134)) and in community discussions on connector API evolution (see [shared doc](https://docs.google.com/document/d/1933jLRluUpQRFLK_w7nJQ9JIUqvoHCXQoJdWLLmox6Q)).

It’s true that **today**, successful **async** KV loading is still reported at the request level. But ideally, the connector should remain request-agnostic and report both successes and failures at the block level, possibly in a progressive and out-of-order manner (for async loading/saving). Mapping blocks back to requests is naturally the scheduler’s responsibility. Looking forward, I think we should optimize this part of the design — for example, introducing a dedicated block-to-request mapping would simplify the scheduler logic.

In the meantime, I’ve tried to encapsulate the error-handling logic into short, focused methods so that readability is maintained and the happy path remains minimally affected.

Commit:
Refactor total_affected_tokens calculation

Signed-off-by: David Ben-David <davidb@pliops.com>

Commit:
Add note on reporting only loadable KV prefix in get_num_new_matched_tokens

Signed-off-by: David Ben-David <davidb@pliops.com>

Commit:
Rename recovered_req_ids to affected_req_ids for clarity

Signed-off-by: David Ben-David <davidb@pliops.com>

njhill commented on examples/offline_inference/kv_load_failure_recovery/rogue_shared_storage_connector.py:
```diff
@@ -0,0 +1,54 @@
+# SPDX-License-Identifier: Apache-2.0
```
@sdavidbd also wondering about this. Do we have equivalent of this also covered in the tests?

njhill commented on vllm/distributed/kv_transfer/kv_connector/v1/base.py:
```diff
@@ -227,6 +227,16 @@ def get_finished(
 
+    def get_block_ids_with_load_errors(self) -> set[int]:
+        """
+        Get the set of block IDs that failed to load.
+
+        Returns:
+            Set of block IDs that encountered load errors.
+            Empty set if no load errors occurred.
```
I think we should be a bit more explicit about the semantics especially for the async case. i.e. failed blocks can be reported for async-loading requests in the steps up to and including the step where the request's id is returned from `get_finished()`; failed async loading requests must still report finished irrespective of failures.

njhill reviewed (commented):
Thanks again for your patience @sdavidbd. It takes me quite a long time to wrap my head around all of the logic in detail.

I agree with @ApostaC, hopefully we can get this merged in the next few days.

Commit:
Refactor KV connector output emptiness check

Signed-off-by: David Ben-David <davidb@pliops.com>

Commit:
Fix PR comments

Signed-off-by: David Ben-David <davidb@pliops.com>

ApostaC reviewed (approved):
Did another on the unit test and examples. Looks like the examples are clear and the tests are sufficient, so LGTM!

njhill reviewed (approved):
Thanks again @sdavidbd for all of your work on this and your extreme patience!

njhill commented on vllm/v1/worker/gpu_input_batch.py:
```diff
@@ -223,6 +228,7 @@ def __init__(
         # NOTE(woosuk): The indices of the requests that do not have their own
         # generator should not be included in the dictionary.
         self.generators: dict[int, torch.Generator] = {}
+        self.generators_last_offset: dict[int, int] = {}
```
Fair enough, we aren't sure how soon this will change so this still might be a nice simplification as a follow-on.

njhill commented:
@sdavidbd a spec decoding test is failing with less-than-expected acceptance length, even after retry. You could try rebasing again incase it's something unrelated already fixed on main...

Commit:
Fix rollback of invalid output tokens and generator state

Signed-off-by: David Ben-David <davidb@pliops.com>

sdavidbd commented:
> @sdavidbd the same test is still failing in the same way, and I verified that it's passing on main. So it seems likely that the changes here are related _somehow_.

Indeed, the regression is on me. The `spec_decode` test failed because resumed requests were being misclassified as having a sync-KV-load failure, which caused their last sampled token to be ignored and reduced the acceptance length.

To disambiguate these cases, the `Scheduler` **must propagate additional information**. My fix is to have the `Scheduler` report the **actual output token count** for each request in `CachedRequestData `within `SchedulerOutput`. This makes it straightforward to detect true sync KV-load failures even when a request has been resumed from preemption, and I find it to be a clear and robust approach.

sdavidbd commented:
> > @sdavidbd the same test is still failing in the same way, and I verified that it's passing on main. So it seems likely that the changes here are related _somehow_.
> 
> Indeed, the regression is on me. The `spec_decode` test failed because resumed requests were being misclassified as having a sync-KV-load failure, which caused their last sampled token to be ignored and reduced the acceptance length.
> 
> To disambiguate these cases, the `Scheduler` **must propagate additional information**. My fix is to have the `Scheduler` report the **actual output token count** for each request in `CachedRequestData `within `SchedulerOutput`. This makes it straightforward to detect true sync KV-load failures even when a request has been resumed from preemption, and I find it to be a clear and robust approach.

@njhill Note: this change will require suppressing BC lint due to the added field in `CachedRequestData` class — I’ll do that once you confirm you’re good with the approach.

njhill commented:
Thanks @sdavidbd, reviewing your latest updates right now.

njhill commented on vllm/v1/worker/gpu_input_batch.py:
```diff
@@ -48,6 +48,13 @@ class CachedRequestState:
         self.num_prompt_tokens = length_from_prompt_token_ids_or_embeds(
             self.prompt_token_ids, self.prompt_embeds)
+        # Establish the initial rollback checkpoint for sync-KV-load failures:
+        # capture the current output-token count (expected to be 0) and, if
```
Are there any cases where it would not be 0?

njhill commented on vllm/v1/worker/gpu_input_batch.py:
```diff
@@ -48,6 +48,13 @@ class CachedRequestState:
+        # capture the current output-token count (expected to be 0) and, if
+        # present, the generator’s internal state. If the request is later
+        # preempted, this checkpoint is refreshed when the request is resumed.
+        self.rollback_num_output_tokens = len(self.output_token_ids)
```
Thinking aloud but wondering whether `original_num_output_tokens` or `starting_num_output_tokens` would be clearer in terms of what the value actually corresponds to.

njhill commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -632,8 +632,36 @@ def _update_states(self, scheduler_output: "SchedulerOutput") -> None:
 
             # Update the cached states.
+
+            if (num_computed_tokens <= req_state.num_computed_tokens and
+                    num_output_tokens == req_state.rollback_num_output_tokens):
+                # Request was rescheduled after a sync-KV-load failure. Remove
+                # any tokens sampled past the rollback checkpoint and restore
```
Would it be better / might it be possible to make indication of this case more explicit? It seems `rollback_num_output_tokens` is only used in an indirect way anyhow (comparing with `num_output_tokens`).

njhill reviewed (commented):
Thanks @sdavidbd, please see inline comments/questions.

sdavidbd commented:
@njhill I’ve simplified the handling here by removing generator rollback from this PR. Rollback of the generator state in the case of sync-KV-load failures can be handled in a follow-up PR, along with unifying and refactoring the rollback mechanism more broadly.

This keeps the logic here focused on correcting state alignment for discarded output tokens. It also directly addresses your comments around overhead, clarity, and duplication without overcomplicating this change.

sdavidbd commented:
The `v1-test-others` failure appears to originate from changes introduced in PR #25896. Build reference: https://buildkite.com/vllm/ci/builds/32924/steps/canvas?sid=019999e7-f4f7-4f46-b184-91f105f6cee4.

njhill commented:
Thanks @sdavidbd that does seem much cleaner.

Could you confirm the implications of not handling the generator rollback?

Seeded outputs may not be consistent in the face of kv load failure/recovery, would that only apply to the sync kv-load-failure case rather than async? (sorry I know I could figure this out but thought I would save time by asking!)

sdavidbd commented:
> Thanks @sdavidbd that does seem much cleaner.
> 
> Could you confirm the implications of not handling the generator rollback?
> 
> Seeded outputs may not be consistent in the face of kv load failure/recovery, would that only apply to the sync kv-load-failure case rather than async? (sorry I know I could figure this out but thought I would save time by asking!)

@njhill Yes, exactly — it would only apply to the sync kv-load-failure case. In the async case, there isn’t a cached state for the request yet.
